Add a pull request template (#1840)
ginkgo-project/ginkgo: .github/pull_request_template.md

# Summary

> <!-- (Madantory) short summary of this PR. -->

Reviewer: ```suggestion
<!-- (Mandatory) short summary of this PR. -->
```

# Details

> <!-- the details in pr like some useful information -->

Reviewer: ```suggestion
<!-- (Optional) more details about the PR. -->
```

# Reminders 

> <!-- for developers and reviewers because most of them are not checked automatically. 

Reviewer: ```suggestion
<!-- For developers and reviewers because most of them are not checked automatically. 
```

Ensure you have set up the pre-commit and the corresponding pre-commit hook.  

> Use `pre-commit run --from-ref origin/develop --to-ref HEAD` to ensure the branche changes compared to origin/develop are formatted through pre-commit.

Reviewer: ```suggestion
Use `pre-commit run --from-ref origin/develop --to-ref HEAD` to ensure the branch changes compared to origin/develop are formatted through pre-commit.
```

Author: Thanks! It is surprising to me that I have so dump typo here 🫢

> Also, check https://github.com/ginkgo-project/ginkgo/wiki/Contributing-guidelines especially about naming-scheme, Whitespace, Other Code Formatting not handled by ClangFormat. 

Reviewer: ```suggestion
Also, check https://github.com/ginkgo-project/ginkgo/wiki/Contributing-guidelines, especially regarding Ginkgo's naming scheme, whitespace guidance, and other code formatting not handled by ClangFormat. 
```

Reviewer: Yeah, it just bothers me that it reads strangely in the text :) Especially the dash in "naming-scheme".

What about adding direct links to each section? E.g. something like

Also, check the [contributing guidelines](https://github.com/ginkgo-project/ginkgo/wiki/Contributing-guidelines), especially regarding the [naming scheme](https://github.com/ginkgo-project/ginkgo/wiki/Contributing-guidelines#naming-scheme), [whitespace use](https://github.com/ginkgo-project/ginkgo/wiki/Contributing-guidelines#whitespace), and [other code formatting not handled by ClangFormat](https://github.com/ginkgo-project/ginkgo/wiki/Contributing-guidelines#other-code-formatting-not-handled-by-clangformat).

Author: because it is not shown in the rich view, it gives several link url in the text

Reviewer: Is that a problem?

I think we should at least be consistent -- right now, "naming-scheme" is the label for the internal link, while "Whitespace" (capitalized) and "Other Code Formatting not handled by ClangFormat" are from the header text (the text from the wiki comes from CONTRIBUTING.md, right?).

What about

```
Also, check https://github.com/ginkgo-project/ginkgo/wiki/Contributing-guidelines, particularly the `Naming scheme`, `Whitespace`, and `Other Code Formatting not handled by ClangFormat` sections.
```

> Please delete the unrelevant entries in the followings.

Reviewer: ```suggestion
Please delete the irrelevant entries in the following.
```

-->
## Creating/Modifying classes 
- [ ] have entries in test/test_install
- [ ] If it is a LinOpFactory, add the corresponding key in `core/config/registry.cpp:generate_config_map`
- [ ] If any changes are introduced in LinOpFactory parameters, add the changes into the corresponding parse function
## Creating/Modifying functions
- [ ] have the corresponding tests
## Introducing new files

> - [ ] ensure they are be added in corresponding CMakeLists.txt

Reviewer: ```suggestion
- [ ] ensure they are added in corresponding CMakeLists.txt
```

## Creating/Modifying kernels
- [ ] have the corresponding entries in core/device_hooks/common_kernels.inc.cpp

> - [ ] check whether the changes are also required for different backend

Reviewer: ```suggestion
- [ ] check whether the changes are also required for other backends
```
